Reject traits with associated types/consts

A trait with associated types fails to compile. Suggest to check for those conditions in impl/src/tagged_trait.rs::expand().

```console
#[typetag::serde(tag = "type")]
trait WebEvent {
    type A;
    fn inspect(&self);
}

/typetag.git $ cargo run --example web_event
   Compiling typetag v0.1.3 (/ws/src/rust/typetag.git)
error[E0191]: the value of the associated type `A` (from the trait `WebEvent`) must be specified
 --> examples/web_event.rs:5:1
  |
5 | #[typetag::serde(tag = "type")]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ associated type `A` must be specified
6 | trait WebEvent {
7 |     type A;
  |     ------- `A` defined here

error: aborting due to previous error
```

--- a/impl/src/parse.rs
+++ b/impl/src/parse.rs
@@ -1,6 +1,6 @@
 use quote::quote;
 use syn::parse::{Parse, ParseStream, Result};
-use syn::{Attribute, Error, ItemImpl, ItemTrait, LitStr, Token, Visibility};
+use syn::{Attribute, Error, ImplItem, ItemImpl, ItemTrait, LitStr, Token, TraitItem, Visibility};
 
 mod kw {
     syn::custom_keyword!(tag);
@@ -33,6 +33,21 @@ impl Parse for Input {
 
         if ahead.peek(Token![trait]) {
             let mut item: ItemTrait = input.parse()?;
+            for assoc in &item.items {
+                if let TraitItem::Const(assoc) = assoc {
+                    let const_token = assoc.const_token;
+                    let semi_token = assoc.semi_token;
+                    let span = quote!(#const_token #semi_token);
+                    let msg = "typetag trait with associated const is not supported yet";
+                    return Err(Error::new_spanned(span, msg));
+                } else if let TraitItem::Type(assoc) = assoc {
+                    let type_token = assoc.type_token;
+                    let semi_token = assoc.semi_token;
+                    let span = quote!(#type_token #semi_token);
+                    let msg = "typetag trait with associated type is not supported yet";
+                    return Err(Error::new_spanned(span, msg));
+                }
+            }
             attrs.extend(item.attrs);
             item.attrs = attrs;
             Ok(Input::Trait(item))
@@ -44,6 +59,21 @@ impl Parse for Input {
                 let span = quote!(#impl_token #ty);
                 let msg = "expected impl Trait for Type";
                 return Err(Error::new_spanned(span, msg));
+            }
+            for assoc in &item.items {
+                if let ImplItem::Const(assoc) = assoc {
+                    let const_token = assoc.const_token;
+                    let semi_token = assoc.semi_token;
+                    let span = quote!(#const_token #semi_token);
+                    let msg = "typetag trait with associated const is not supported yet";
+                    return Err(Error::new_spanned(span, msg));
+                } else if let ImplItem::Type(assoc) = assoc {
+                    let type_token = assoc.type_token;
+                    let semi_token = assoc.semi_token;
+                    let span = quote!(#type_token #semi_token);
+                    let msg = "typetag trait with associated type is not supported yet";
+                    return Err(Error::new_spanned(span, msg));
+                }
             }
             attrs.extend(item.attrs);
             item.attrs = attrs;

--- a/tests/ui/associated-const.stderr
+++ b/tests/ui/associated-const.stderr
@@ -1,15 +1,5 @@
-error[E0038]: the trait `Trait` cannot be made into an object
- --> tests/ui/associated-const.rs:1:1
+error: typetag trait with associated const is not supported yet
+ --> tests/ui/associated-const.rs:3:5
   |
-1 | #[typetag::serde]
-  | ^^^^^^^^^^^^^^^^^ `Trait` cannot be made into an object
-  |
-note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
- --> tests/ui/associated-const.rs:3:11
-  |
-2 | pub trait Trait {
-  |           ----- this trait cannot be made into an object...
 3 |     const ASSOC: u8;
-  |           ^^^^^ ...because it contains this associated `const`
-  = help: consider moving `ASSOC` to another trait
-  = note: this error originates in the attribute macro `typetag::serde` (in Nightly builds, run with -Z macro-backtrace for more info)
+  |     ^^^^^^^^^^^^^^^^

--- a/tests/ui/associated-type.stderr
+++ b/tests/ui/associated-type.stderr
@@ -1,7 +1,5 @@
-error[E0191]: the value of the associated type `Assoc` (from trait `Trait`) must be specified
- --> tests/ui/associated-type.rs:2:11
+error: typetag trait with associated type is not supported yet
+ --> tests/ui/associated-type.rs:3:5
   |
-2 | pub trait Trait {
-  |           ^^^^^ help: specify the associated type: `Trait<Assoc = Type>`
 3 |     type Assoc;
-  |     ---------- `Assoc` defined here
+  |     ^^^^^^^^^^^

